### ISSUE TITLE
self.model has no attribute 'in_features'

### ISSUE THREAD

**Issue by lhj940825** [asharakeh/pod_compare#6]
First of all, thank you for sharing your code. 

I would like to ask your ideas for two issues I have met.
**1.** While I was working on running your evaluation (apply_net.py) with the method: **BayesOD+Dropout**, I have encountered the following error. 
![image](https://user-images.githubusercontent.com/47107292/119550734-09faa800-bd99-11eb-87da-5335bd595632.png)
It says self.model has no attribute named '**in_features**'. While debugging, I found that self.model has an attribute '**head_in_features**' instead. Is it okay to replace 'in_features' by 'head_in_features' ?

The **command** I used is: 
`python src/apply_net.py --dataset-dir /My/BDD/Dir --test-dataset bdd_val --config-file BDD-Detection/retinanet/retinanet_R_50_FPN_1x_reg_cls_var_dropout.yaml --inference-config Inference/bayes_od_mc_dropout.yaml
` with your pre-trained model(`retinanet_R_50_FPN_1x_reg_cls_var_dropout.pth`)

**2.** When I run the **command** above, I get the notification as below:
![image](https://user-images.githubusercontent.com/47107292/119553121-9c03b000-bd9b-11eb-9868-12fb07a39a53.png)
It states that the checkpoint contains some non-used keys or missing keys. Does this indicate that I am using wrong checkpoint file?
The method I wanted to evaluate is  **BayesOD+Dropout**, and I used the checkpoint (`retinanet_R_50_FPN_1x_reg_cls_var_dropout.pth`).

Thank you for reading.

**Comment by asharakeh** [comment]
Hi @lhj940825, these errors originate from the code not being compatible with the latest detectron2 updates. I will have to spend some time fixing these issues, which I, unfortunately, do not currently have. 

 **Is it okay to replace 'in_features' by 'head_in_features' ?** Yes, that is the correct fix.
**Does this indicate that I am using wrong checkpoint file?** No, I think detectron2 also changed the naming conventions In retinanet model. You might have to retrain the model unfortunately (or map the old parameter names to the new ones).

**Comment by lhj940825** [comment]
Hi, thank you for the reply. 

After your response, I have tried to run the code with the virtual env, which is installed as is described in the README, but the issues still remain (not existing self.model.in_features, missing keys in checkpoint file).

Also when I replace self.model.in_features by **self.model.head_in_features**, I encounter another error shown below
![image](https://user-images.githubusercontent.com/47107292/119965497-9fb95180-bfaa-11eb-944a-5f6919258a7e.png)


It seems still I did not perfectly reproduce the virtual env you used. 
Could you share the requirement.txt stating the exact versions of libraries you used to construct the virtual env?

**Comment by asharakeh** [comment]
Did you try setting >= to == in https://github.com/asharakeh/pod_compare/blob/main/requirements.txt ?

To be honest, I have not ran this code since November 2020, but I believe the lower bound versions in the requirements should be the ones we used to run the code at the time.

The error seems to stem from new version of pytorch. This repo supports versions up to 1.6 only.

**Comment by lhj940825** [comment]
>Did you try setting >= to == in https://github.com/asharakeh/pod_compare/blob/main/requirements.txt ?

instead of trying this, I simply train your model from scratch, and it seems to work. 
However, after around 40K iterations, the loss_box_reg becomes negative, yielding a negative total loss. 

![image](https://user-images.githubusercontent.com/47107292/120552836-08b02780-c3f8-11eb-983e-2cbc3c5a3647.png)
Is this a normal behavior of the loss when training model with the **method: BayesOD+Dropout**?

**Comment by asharakeh** [comment]
The negative log-likelihood loss ranges between -infinity and + infinity, so yes it can take negative values. The values seem reasonable.
